### PR TITLE
`after_trial` implemented for `TPESampler`

### PR DESCRIPTION
**Pull request by jeromepatel**
## Motivation
With reference to the issue #2233, and continuation of work PR #2372 I have implemented <code>after_trial</code> method for <code>TPESampler</code> with the corresponding testing function <code>test_call_after_trial_of_independent_sampler</code>

## Description of the changes
 As the <code>independent_sampler</code> didn't exist of now, but <code>random_sampler</code> was there. So, I have implemented <code>after_trial</code> for <code>_random_sampler</code>, the corresponding changes for tests are also included in <code>tests/samplers_tests/tpe_tests/test_sampler.py</code>. 
 ### Samplers

- [x]  `PartialFixedSampler` #2209
- [x]   `CmaEsSampler` #2359 
- [x]   `PyCmaEsampler` #2365 
- [x]  `SkoptSampler` #2372 
- [x]   `BoTorchSampler` #2372 
- [x]  `TPESampler` (this PR)
- [ ]   `NSAGAIISampler`
- [ ]  `MOTPESampler`

If this implementation needs any further modification then let me know.
Thank you!!

### CODE REVIEW

**Comment by 0x41head** [comment]
@jeromepatel you seem to have referenced the wrong pull request in your checklist.
The one you wanted to reference was #2372 , however, you have referenced my PR #2375 which is unrelated to this PR.

**Comment by jeromepatel** [comment]
I am extremely sorry for that. I will correct it. Thank you for informing

**Comment by 0x41head** [comment]
It's alright. Just wanted to let you know 👍

**Comment by toshihikoyanase** [comment]
@hvy @HideakiImamura Could you review this PR?

I personally feel that introducing the `independent_sampler` argument to `TPESampler.__init__` can be out-of-scope of the issue #2233, but I'm not fully sure. Could you clarify the scope before diving into the details?

**Comment by jeromepatel** [comment]
Hello @toshihikoyanase, the <code>TPESampler</code> needed an <code>after_trial</code> method, as all the other previous samplers had <code>independent_sampler</code> variable preexisting, it was easy to introduce <code>after_trial</code>.However for <code>TPESampler</code> there is <code>_random_sampler</code> which is hard-coded or initialized with <code>RandomSampler</code> and used further. I have changed only the name to <code>independent_sampler</code> (by default initialized with <code>RandomSampler</code>), thus the parameter <code>independent_sampler</code> provides flexibility to users which sampler to use when a fallback occurs for <code>TPESampler</code>.

**Comment by hvy** [comment]
I agree that giving the user flexibility to swap the wrapped sampler for TPE could be useful (especially when we introduce other Quasi-Monte Carlo samplers such as Sobol). However we could separate this discussion from the trial post processing with `after_trial`.

Although I'm neutral to this, I know that there are opinions that we don't need to implement `after_trial` for samplers that don't need them in practice (e.g. don't allow arbitrary wrapped trials).

**Comment by jeromepatel** [comment]
So, to have <code>after_trial</code> method for <code>TPESampler</code> separate and clean from <code>independent_sampler</code> disuccsion and PR, should I revert the name <code>independent_sampler</code> with <code>random_sampler</code> added with just <code>after_trial</code> method ?

**Comment by toshihikoyanase** [comment]
@jeromepatel Thank you for your replies. I understand the reason why you introduced the argument. As @hvy mentioned, users may want to switch the random sampler after Sobol sampler (#1797) is merged. Let us discuss it in a new issue/PR.

**Comment by jeromepatel** [comment]
Thank you for your guidance @hvy  and @toshihikoyanase, I will replace <code>independent_sampler</code> with <code>random_sampler</code>, this PR should include only <code>after_trial</code>change and I'll add a new issue for proposal of <code>independent_sampler</code> in <code>TPESampler</code>.

**Comment by jeromepatel** [comment]
@hvy I have modified this PR files ( <code>tests/samplers_tests/tpe_tests/test_sampler.py</code> and <code>optuna/samplers/_tpe/sampler.py</code>) for implementation of only <code>after_trial</code> method, the changes made for <code>independent_sampler</code> were discarded for this PR. I will create a new issue on discussion of adding this new parameter to <code>TPESampler</code>. Let me know if it needs any further modification.

**Comment by HideakiImamura** [comment]
@jeromepatel Thanks for the PR. IMO, the name of `independent_sampler` (or `random_sampler`) does not matter. I think we should discuss introducing `independent_sampler` at first. After the discussion, if we introduce the argument, we should revisit this PR to call `after_trial` of the independent sampler.

**Comment by jeromepatel** [comment]
@HideakiImamura I agree with you, I have discarded the parameter <code>independent_sampler</code> as of now and continuing with <code>random_sampler</code>(which was already present in <code>TPESampler</code>). I am opening a new issue to discuss whether <code>independent_sampler</code> should be added or not. After that, I can implement that based on the decision.  Thank you!

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 1]
```diff
@@ -869,3 +869,15 @@ def test_reseed_rng() -> None:
         sampler.reseed_rng()
         assert mock_object.call_count == 1
         assert original_seed != sampler._rng.seed
+
+
+def test_call_after_trial_of_independent_sampler() -> None:
```
Comment (HideakiImamura) [inline tests/samplers_tests/tpe_tests/test_sampler.py:874]:
```suggestion
def test_call_after_trial_of_random_sampler() -> None:
```

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 2]
```diff
@@ -869,3 +869,15 @@ def test_reseed_rng() -> None:
...
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", optuna.exceptions.ExperimentalWarning)
+        sampler = TPESampler()
```
Comment (HideakiImamura) [inline tests/samplers_tests/tpe_tests/test_sampler.py:877]: `TPESampler` is not experimental.
```suggestion
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have minor two comments. PTAL.

File: tests/samplers_tests/tpe_tests/test_sampler.py (hunk 1 above)
Comment (jeromepatel) [reply tests/samplers_tests/tpe_tests/test_sampler.py:874]: Okay, I'll change that.

File: tests/samplers_tests/tpe_tests/test_sampler.py (hunk 2 above)
Comment (jeromepatel) [reply tests/samplers_tests/tpe_tests/test_sampler.py:877]: I'll remove that. Thank you for reviewing.

**Review by HideakiImamura — APPROVED** [review]
Thanks for your swift actions! LGTM!

**Review by hvy — APPROVED** [review]
Thanks as always, LGTM!

**Pull request merged**